feat: add a decimals parameter, and decimals method to brand

katelynsills opened the pull request:
This PR adds a `displayInfo` object to an ERTP issuerKit, including a `decimalPlaces` property in the same way that Ethereum uses their `decimals` property. We intend for fungible digital assets to be represented as integers in the smallest unit. (In finance, this is often [mills](https://en.wikipedia.org/wiki/Mill_(currency)).) However, we want to display in a more user-friendly way. `decimalPlaces` is meant **only for display**, and informs the UI how many decimal places to move the decimal over to the left. For instance, for mills, one thousandth of a dollar, decimals would be 3.

The downside is that this hardcodes a certain interpretation. For instance, it would hard-code a dollar as the standard representation instead of millions or thousands of dollars. However, the UI is free to ignore `decimals`. 

One upside is that compared to something like a custom `toString`, decimals is very safe. An attacker can't do much with the ability to choose an integer that will later be returned.

Closes #119

michaelfig reviewed (approved):
LGTM!

The brand is a good place to put this information, in case amountMath is later sensitive to it.

michaelfig commented on packages/ERTP/src/issuer.js:
```diff
@@ -14,11 +14,9 @@ import { makeInterface, ERTPKind } from './interfaces';
+ * @type {MakeIssuerKit}
  */
-function makeIssuerKit(allegedName, amountMathKind = MathKind.NAT) {
+function makeIssuerKit(allegedName, amountMathKind = MathKind.NAT, decimals) {
```
```suggestion
function makeIssuerKit(allegedName, amountMathKind = MathKind.NAT, decimals = undefined) {
```

katelynsills replied:
This was frowned upon by the typescript enforcement because number cannot be undefined. I actually had this at first and had to revert.

erights commented on packages/ERTP/src/issuer.js:
```diff
@@ -14,11 +14,9 @@ import { makeInterface, ERTPKind } from './interfaces';
+ * @type {MakeIssuerKit}
  */
-function makeIssuerKit(allegedName, amountMathKind = MathKind.NAT) {
+function makeIssuerKit(allegedName, amountMathKind = MathKind.NAT, decimals) {
```
This is weirdly special case, to include a parameter whose name makes it specific to one particular choice of the previous parameter, `nat`, when that previous parameter represents an (eventually) extensible set of choices. What happens when we have another math kind whose meaning is conditioned on other parameters?

```suggestion
function makeIssuerKit(allegedName, amountMathKind = MathKind.NAT, options = {}) {
  const { decimalPlaces } = options;
```

The meaning of `decimal` is unclear. When I talk about this verbally, I say "decimal places".

katelynsills replied:
Yeah, I agree. Ok, I will adopt this! Instead of `options`, what if I call this `displayData` or `displayInfo`? If later, we want to expand it, we can do so without causing incompatibilities since the name is only at the documentation level.

Chris-Hibbert commented on packages/ERTP/src/issuer.js:
```diff
@@ -32,6 +30,12 @@ function makeIssuerKit(allegedName, amountMathKind = MathKind.NAT) {
+      // Fungible digital assets should be represented in integers, in
+      // the smallest unit (i.e. USD might be represented in mill,
+      // a thousandth of a dollar. In that case, `decimals` would be 3).
+      decimals: () => decimals,
```
How about 'presentationHints' or something in that direction?

katelynsills replied:
how about `getDisplayData`?

Chris-Hibbert reviewed (commented):
> The downside is that this hardcodes a certain interpretation. 

I think the right way to think of this is that 'decimals' tells the display software which position corresponds to whole numbers. If the display wants to display Millions, then it knows that it should be millions of that unit, rather than millions of the most precise division.

if mills are the unit, and decimals is set to 3, then 
 * 3000 is $3.00
 * 3 000 000 is $3000.00 
 * 3 000 000 000 could be rendered $3.0M

katelynsills commented:
Addressed PR comments. In order to ensure `displayInfo` is as expected, I copied over a few helpers from Zoe. Not happy about that, but I also don't want to create a package for two helpers before the hackathon.

erights commented on packages/ERTP/src/displayInfo.js:
```diff
@@ -0,0 +1,38 @@
+  );
+};
+
+export const assertDisplayInfo = allegedDisplayInfo => {
```
There are a number of ways that `allegedDisplayInfo` may misbehave that this code does not protect against. The simplest and safest defense would be to first make a `pureCopy` using https://github.com/Agoric/agoric-sdk/blob/76e0723915a32b6b76ebea55e4df29d67f8afea8/packages/marshal/src/marshal.js#L30-L32

erights commented on packages/ERTP/src/issuer.js:
```diff
@@ -10,14 +10,20 @@ import { isPromise } from '@agoric/promise-kit';
+  displayInfo = undefined,
+) {
   assert.typeof(allegedName, 'string');
+  assertDisplayInfo(displayInfo);
```
There are a number of ways that `allegedDisplayInfo` may misbehave that this code does not protect against. The simplest and safest defense is to *first* make a `pureCopy` using https://github.com/Agoric/agoric-sdk/blob/76e0723915a32b6b76ebea55e4df29d67f8afea8/packages/marshal/src/marshal.js#L30-L32

```suggestion
  displayInfo = pureCopy(displayInfo);
  assertDisplayInfo(displayInfo);
```

katelynsills replied:
Done!

erights commented on packages/ERTP/src/types.js:
```diff
@@ -97,6 +97,19 @@
+ *   numbers. We require fungible digital assets to be represented in
+ *   integers, in the smallest unit (i.e. USD might be represented in mill,
+ *   a thousandth of a dollar. In that case, `decimalPlaces` would be 3.)
+ *   For non-fungible digital assets, this should be left as undefined.
```
`DisplayInfo` should be expected to grow over time as one would expect of an options bag. Each use would only contain properties appropriate for the kind of amount math it is used on. Rather than having `decimalPlaces` be `undefined`, the property should be optional. AFAICT your `assertDisplayInfo` already allows it to be absent so it is just an issue of how you type and describe it here.

katelynsills replied:
ah, yep, that is what I intended. Will change

erights commented on packages/ERTP/src/types.js:
```diff
@@ -97,6 +97,19 @@
+ *   integers, in the smallest unit (i.e. USD might be represented in mill,
+ *   a thousandth of a dollar. In that case, `decimalPlaces` would be 3.)
+ *   For non-fungible digital assets, this should be left as undefined.
+ *   The decimalPlaces property should be used for *display purposes only*. Any
+ *   other use is an anti-pattern.
```
Nice caution! 

Btw, JS Internationalization has a bunch of stuff about formatting currency. Those formatters are also only about rendering things into strings, not the underlying representation. I can imagine we might eventually want to grow this options bag in that direction. But not yet ;)

katelynsills replied:
Agreed!

erights reviewed (changes requested):
Once the safety issue is addressed with the `pureCopy`, LGTM.

erights commented on packages/ERTP/src/issuer.js:
```diff
@@ -23,7 +23,7 @@ function makeIssuerKit(
 ) {
   assert.typeof(allegedName, 'string');
-  assertDisplayInfo(displayInfo);
+  displayInfo = coerceDisplayInfo(displayInfo);
```
Good you changed the assert into a coerce and let it do the coercion. Better than what I suggested.

erights commented on packages/ERTP/src/displayInfo.js:
```diff
@@ -40,3 +41,12 @@ export const assertDisplayInfo = allegedDisplayInfo => {
+
+export const coerceDisplayInfo = allegedDisplayInfo => {
+  if (passStyleOf(allegedDisplayInfo) === REMOTE_STYLE) {
+    return harden({});
```
This is too loose. Once you see it is a remotable, you should also assert that it is an empty object before returning a new empty object. Otherwise, an accidental passing of a substantive remotable object will silently become an empty object rather than rejected with a nice diagnostic.

katelynsills replied:
@erights How would you assert it is an empty object if it's a presence? I thought doing something like getting the ownPropertyNames wouldn't work.

erights replied:
See #1981

erights reviewed (approved):
LGTM
